### DISCUSSION TITLE
update child view from parent does not work when state change

### DISCUSSION THREAD

**Discussion by saroar** [pointfreeco/swift-identified-collections#11] [General]
```swift

public struct ConversationsState: Equatable {
  ..........
  public var conversations: IdentifiedArrayOf<ConversationResponse.Item> = []
}

public struct TabsState: Equatable {
  ........
  public var conversations: ConversationsState
}

public let tabsReducer = Reducer<TabsState, TabsAction, TabsEnvironment> { state, action, environment in
 ..............................
  case let .receivedSocketMessage(.success(.string(str))):

      guard let data = str.data(using: .utf8) else {
        return .none
      }

      let chatOutGoingEvent = ChatOutGoingEvent.decode(data: data)

      switch chatOutGoingEvent {
      case .conversation(let message):

          // before update working update child my view
          guard let index = state.conversations.conversations
            .firstIndex(where: { $0.id == message.conversationId }) else {
          return .none
          }
          state.conversations.conversations[index].lastMessage = message
          // old code crash becz
         // "use the id-based subscript, instead, for in-place modification"

          // after which not working not update my child view
          state.conversations.conversations[id: message.conversationId]?
          .lastMessage = message
        return .none
      case .message(let message):
        state.conversations.chatState?.messages.insert(message, at: 0)
        return .none
      }
}

```

**Comment by mbrandonw** [comment]
I am going to convert this to a discussion as it doesn’t appear to be an issue with the library.

**Reply by saroar** [reply]
if it's not working as expected where is the problem?
how I can find which library having this issue?

**Reply by mbrandonw** [reply]
It is most likely a problem with your reducer, and so you could open a [discussion](http://github.com/pointfreeco/swift-composable-architecture/discussions) on the Composable Architecture repo. If you do that, then please provide more information about what behavior you are seeing and what behavior you expect. It would also be really helpful to have some code that can be run.

**Reply by saroar** [reply]
I have an open-source app but there is not test is written only having problem after update TCA

**Reply by mbrandonw** [reply]
Oh, so this is an issue that came up with the 0.21 release but wasn't happening in the previous release?

**Reply by saroar** [reply]
yes becz in my last release 

here i was add little code 

```swift
public struct ConversationsState: Equatable {
  ..........
  public var conversations: IdentifiedArrayOf<ConversationResponse.Item> = []
}

public struct TabsState: Equatable {
  ........
  public var conversations: ConversationsState
}

public let tabsReducer = Reducer<TabsState, TabsAction, TabsEnvironment> { state, action, environment in
 ..............................
  case let .receivedSocketMessage(.success(.string(str))):

      guard let data = str.data(using: .utf8) else {
        return .none
      }

      let chatOutGoingEvent = ChatOutGoingEvent.decode(data: data)

      switch chatOutGoingEvent {
      case .conversation(let message):

          // before update working update child my view
          guard let index = state.conversations.conversations
            .firstIndex(where: { $0.id == message.conversationId }) else {
          return .none
          }
          state.conversations.conversations[index].lastMessage = message
          // old code crash becz
         // "use the id-based subscript, instead, for in-place modification"

          // after which not working not update my child view
          state.conversations.conversations[id: message.conversationId]?
          .lastMessage = message
        return .none
      case .message(let message):
        state.conversations.chatState?.messages.insert(message, at: 0)
        return .none
      }
}
```

**Reply by saroar** [reply]
u can check the old version code here all was fine 
https://github.com/AddaMeSPB/AddaMeIOS/blob/feature/EventDetaulsUI%2BChatUI/Addame/Addame/Sources/TabsView/TabsReducer.swift#L173

**Reply by stephencelis** [reply]
@saroar We dropped `MutableCollection` and `RangeReplaceableCollection` conformances in this release, since they contain methods that are unsafe and can break invariants. Instead of plucking out the index for mutation, you can mutate via the `[id:]` subscript:

```diff
-if let row = state.conversations.conversations.firstIndex(where: { $0.id == message.conversationId }) {
-  state.conversations.conversations[row].lastMessage = message
-}
+state.conversations.conversations[id: message.conversationId]?.lastMessage = message
```

**Reply by saroar** [reply]
i think you dont check my code i update from old to new then ui never update 

switch chatOutGoingEvent {
      case .conversation(let message):

          // before update working update child my view
          guard let index = state.conversations.conversations
            .firstIndex(where: { $0.id == message.conversationId }) else {
          return .none
          }
          state.conversations.conversations[index].lastMessage = message
          // old code crash becz
         // "use the id-based subscript, instead, for in-place modification"

          // after which not working not update my child view
          state.conversations.conversations[id: message.conversationId]?
          .lastMessage = message
        return .none
@stephencelis :( i missed this code u see i write before and after

**Reply by stephencelis** [reply]
@saroar Ah yeah, I didn't see that part of the code sample. Can you explain what you mean by "not working"? I tried checking out the code and it seems to compile.

**Reply by saroar** [reply]
when some will write something in this conversation room I get msg from WebSocket then update my last message and UI will move to the top of the list and update the last message 
![IMG_AD3588F0FF72-1](https://user-images.githubusercontent.com/8770772/125969729-b27153d3-ad77-4502-a732-1f0b5c24602a.jpeg)

**Reply by stephencelis** [reply]
Which of the following are you saying?

1. Before: Conversation updated in place.
    After: Conversation updates and moves to top.

2. Before: Conversation updates and moves to top.
    After: Conversation updated in place.

3. Something else?

Do you think it's possible to create a smaller reproduction and explain exactly what the change in behavior is (what it was before, what it is now, what you expect it to be, etc.)? Sorry if there's a communication barrier here, we're just having a hard time understanding the problem.

**Reply by saroar** [reply]
1 is correct although it’s very simple main goal is update UI after state change  before update my tca working find as expected my open source link already there can’t create any app becz it lots of work becz I need to handle websocket etc

**Reply by saroar** [reply]
Tca 0.20.0 working fine with index update this way
state.conversations.conversations[index].lastMessage = message

**Reply by stephencelis** [reply]
@saroar We'd love to help, but we really need a sample project where we can observe this behavior. Can you please attach a small project that reproduces the problem? Unfortunately I'm unable to get your AddaMe project to build in Xcode, though even if I could it's quite a big project so I'm not sure I'd know how to track down the issue 😅

**Reply by saroar** [reply]
For now, I will use the older version TCA 0.20.0 
only i am having issue is when parent update child state view was not update 
Thanks for try to help @stephencelis

**Reply by stephencelis** [reply]
Sounds good! If you are able to provide a repro in the future, please do share it! Or hopefully someone else will encounter the problem and can help us track down the issue, as well.

**Reply by saroar** [reply]
before i use let = array: [ConversationResponse.Item] = []
but now I am using IdentifiedArrayOf <ConversationResponse.Item> = [] may be I will debug more

**Reply by saroar** [reply]
@stephencelis this how i fixed 
```swift
state.conversations.conversations[id: message.conversationId]? .lastMessage = message
state.conversations.conversations.sort() // <- this line are fixed my issue 
```
may be its dispatch issue
